#  Assignment 5

This assignment covers material on matrices, norms, condition numbers and eigensystems. The assignment is due by 5pm on Friday the 17th of April.

The breakdown of the marks is as follows:

Question 0 - [1 mark]  
Question 1 - [7 marks]  
Questino 2 - [6 marks]  
Question 3 - [7 marks]  
Question 4 - [9 marks]  

# Question 0

You should have saved this notebook to your ACM20030-Assignments folder. Now:
- commit the notebook to your local repository
- push the changes to the GitHub

In [1]:
import numpy as np
from numpy import linalg as la
np.set_printoptions(linewidth=200)  # This stops line wrapping of larger matrices

# Question 1 - practice manipulating arrays

This questions asks you write a few functions for manipulating arrays of data

## Q1 a

Write a function that checks if a given $n\times n$ matrix is symmetric. Use only loops and `.shape` function from numpy (do not use any other special numpy functions).

In [2]:
def checkSymmetric(a):
    n=len(a)
    #print(n)
    for i in range(0,n):
        for j in range(i+1,n):
            #print(i,j,a[i,j],a[j,i])
            if(not a[i,j]==a[j,i]):
                return False
                
    return True

Test your function on the two matrices below

In [3]:
A = np.random.rand(5,5)
SymMatrix = (A + A.T)/2

In [4]:
#print(A)
checkSymmetric(A)

False

In [5]:
checkSymmetric(SymMatrix)

True

## Q1 b

Write a function to transpose a general $n\times n$ matrix. Do this by writing loops and not using any special functions like `np.transpose()`, `A.T`, or any functions from any other library.

In [6]:
def matrixTranspose(a):
    n=len(a)
    out = np.zeros((n,n))
    for i in range(n):
        for j in range(n):
            out[j,i]=a[i,j]
    #print(out)
    return out

def matrixEquals(a,b):
    if (not len(a)==len(b)):
        return False
    if (not len(a[0])==len(b[0])):
        return False
    for i in range(len(a)):
        for j in range(len(a[0])):
            if(not a[i,j]==b[i,j]):
                return False
    return True

Check your code by transposing the below random matrix and checking that $A^T \neq A$ and $(A^T)^T - A = 0$

In [7]:
A = np.random.rand(6,6)
At = matrixTranspose(A)
print("A:",A)
print("A^T:",At)
print("A==A^T?",matrixEquals(A,At))
print("(A^T)^T-A==0?",matrixEquals(matrixTranspose(At)-A,np.zeros((len(A),len(A)))))


A: [[0.7946158  0.45363754 0.3559446  0.90097908 0.40041734 0.75895411]
 [0.65227272 0.13672308 0.38253198 0.86448285 0.32533951 0.82939772]
 [0.00466846 0.84501212 0.8048021  0.85635391 0.90699085 0.17829562]
 [0.06366871 0.52978663 0.55674597 0.53043422 0.23981077 0.21270221]
 [0.58889485 0.16932974 0.91382951 0.67648445 0.51902493 0.09170533]
 [0.00904789 0.19903732 0.26011409 0.57227172 0.33430065 0.51740359]]
A^T: [[0.7946158  0.65227272 0.00466846 0.06366871 0.58889485 0.00904789]
 [0.45363754 0.13672308 0.84501212 0.52978663 0.16932974 0.19903732]
 [0.3559446  0.38253198 0.8048021  0.55674597 0.91382951 0.26011409]
 [0.90097908 0.86448285 0.85635391 0.53043422 0.67648445 0.57227172]
 [0.40041734 0.32533951 0.90699085 0.23981077 0.51902493 0.33430065]
 [0.75895411 0.82939772 0.17829562 0.21270221 0.09170533 0.51740359]]
A==A^T? False
(A^T)^T-A==0? True


## Q1c

Write a function that finds the maximum entry in a matrix (where the matrix does not need to be square). Do this by writing loops and not using any special functions like `np.max` or any functions from any other library.

In [8]:
def matrixMax(a):
    
    maxIndex = (0,0)
    maxValue = a[maxIndex]
    #print(maxValue)
    for i in range(len(a)):
        for j in range(len(a[0])):
            if(a[i,j]>maxValue):
                maxValue = a[i,j]
                maxIndex = (i,j)
    return maxValue

Check your result by comparing to `np.max(A)` for the below random matrix

In [9]:
A = np.random.rand(4,5)
#print(A[0,0])
matrixMax(A)

0.9002438560371514

# Question 2:  matrix norms

In the lectures we focused on the $L^2$-norm. We also briefly mentioned the $L^1$ and $L^\infty$ norm. Let's look at calculating those as well for a general $m\times n$ matrix.

# Q2a

The $L^1$ norm is given by

$$ \lVert A \rVert_1 = \max_j \sum^m_{i=1} |a_{ij}| $$

i.e., it is the max of the sums of the absolute values of each column. Write a function that computes the $L^1$-norm using only loops. The function should use no special numpy functions except `np.abs()`.

In [10]:
def L1Norm(a):
    maxValue = 0
    maxbool=True
    for j in range(len(a[0])):
        currentSum = 0
        for i in range(len(a)):
            currentSum +=np.abs(a[i,j])
        if(maxbool):
            maxbool=False
            maxValue=currentSum
        else:
            if currentSum>maxValue:
                maxValue = currentSum
    return maxValue

Check your result against NumPy's `la.norm(A,1)` function for the following random matrix

In [11]:
A = np.random.rand(3,4)
print("My value for the L^1-norm:",L1Norm(A))
print("NumPy's L^1-norm:",la.norm(A,1))

My value for the L^1-norm: 2.2346650204409473
NumPy's L^1-norm: 2.2346650204409473


## Q2b

The $L^\infty$ norm is given by 

$$\lVert A \rVert_\infty = \max_i \sum_{j=1}^n |a_{ij}|$$

i.e., it is the max of the sums of the absolute values of each row. Write a function that computes the $L^\infty$-norm using only loops and no special numpy functions.

In [12]:
def LInfNorm(a):
    maxValue = 0
    maxbool=True
    for i in range(len(a)):
        currentSum = 0
        for j in range(len(a[0])):
            currentSum +=np.abs(a[i,j])
        if(maxbool):
            maxbool=False
            maxValue=currentSum
        else:
            if currentSum>maxValue:
                maxValue = currentSum
    return maxValue

Check your result against NumPy's `la.norm(A,np.inf)` function for the following random matrix

In [13]:
A = np.random.rand(3,4)
print("My value for the L^inf-norm:",LInfNorm(A))
print("NumPy's L^inf-norm:",la.norm(A,np.inf))

My value for the L^inf-norm: 2.167630990907737
NumPy's L^inf-norm: 2.167630990907737


# Question 3: ill-conditions systems of equations

In this question you can use any of NumPy's functions.

## Q3a

Which of the following three matrices, $A1, A2, A3$, are ill-conditioned. State the reason for your conclusion.

In [14]:
A1 = np.array([[0.988746460106, 0.351486109477, 0.024911477532, 0.384039749683],
       [0.282104657306, 0.239884151775, 0.36170151991 , 0.903082266861],
       [0.59333290915 , 0.637178863968, 0.241229214635, 0.685595037901],
       [0.013591747828, 0.917926970364, 0.458689093885, 0.157821121254]])

A2 = np.array([[ 2.852416473483e-01,  3.803483118694e-01,  2.697704226632e-01,  6.810611849029e-01],
       [ 6.622827828376e-01,  6.447943415836e-01,  4.387842058829e-01,  1.530575702376e+00],
       [ 3.672130860703e-01,  8.309522079704e-01,  9.975487474007e-01,  5.678296337101e-01],
       [ 4.036490000599e-01,  6.865978225967e-06,  8.788095913773e-01, -7.150472427928e-02]])

A3 = np.array([[0.131923621844, 0.245316154195, 0.732362341613, 0.957047891764],
       [0.91290446184 , 0.512312007959, 0.391938831444, 0.092509684772],
       [0.648708523652, 0.608576675797, 0.576773587276, 0.065385122929],
       [0.502955480393, 0.084082960221, 0.659487390444, 0.662958739412]])

In [15]:
print("L^2 condition number of A1",la.cond(A1,2))
print("L^2 condition number of A2",la.cond(A2,2))
print("L^2 condition number of A3",la.cond(A3,2))


L^2 condition number of A1 21.55246648838184
L^2 condition number of A2 13334945724.122156
L^2 condition number of A3 15.122449699410616


A2 has a large condition number ($\simeq 13334945724$), making it ill-conditioned.
A1 and A3 have low condition numbers, so they are not ill-conditioned

# Q3b

With $b = [1,2,3,4]$ and $\delta b = [0.1,0,0,0]$ for $A2$ and $A3$ calculate the percentage change in $x$ due to a small change in $b$, i.e., calculate 

$$100\times\frac{\lVert \delta x_i\rVert_2}{\lVert x_i \rVert_2}\quad\text{and}\quad 100\times\frac{\lVert \delta b\rVert_2}{\lVert b \rVert_2}$$

where $(x_i+\delta x_i)  = A_i^{-1}(b+\delta b)$ and $i\in\{2,3\}$.

In [16]:
b=np.array([1,2,3,4])
db=np.array([.1,0,0,0])

#la.norm(db,2)/la.norm(b,2)
b_per = 100*la.norm(db,2)/la.norm(b,2)
#la.norm(,2)/la.norm(b,2)

x2 = np.dot(la.inv(A2),(b))
x2_dash = np.dot(la.inv(A2),(b+db))
dx2 = x2_dash-x2
x2_per = 100*la.norm(dx2,2)/la.norm(x2,2)

x3 = np.dot(la.inv(A3),(b))
x3_dash = np.dot(la.inv(A3),(b+db))
dx3 = x3_dash-x3
x3_per = 100*la.norm(dx3,2)/la.norm(x3,2)


#np.transpose(A2)*(b+db)

print("change in b(%%): %0.3f%%" % b_per)
print("change in x_2(%%):%0.3f%%" % x2_per)
print("change in x_3(%%):%0.3f%%" % x3_per)

change in b(%): 1.826%
change in x_2(%):512.621%
change in x_3(%):2.174%


Which system of equations is more sensitive to small changes in $b$? How could you have guessed this from the answer to the Q3a?

Answer here

A2 is more sensitive to changes in b.

We could have guessed this from A2's large condition number

# Question 4: eigensystems

In this question we look at finding eigenvalues and eigenvectors

## Q4a

You may use any of NumPy's functions for this sub-question.

Find the eigenvalues and eigenvectors for the below matrix. Print out the eigenvalues sorted by descending value and explicitly show that eigenvalues/eigenvector pairs satisfy $Ax = \lambda x$.

In [17]:
A = np.array([[1,4,5],[4,5,4],[6,7,9]])

In [18]:
eigs = la.eig(A)
#print(eigs)

values = eigs[0]
vectors = eigs[1]

order = []
remove = [0,1,2]
for i in remove:
    largestIn = remove[0]
    largestVal = values[remove[0]]
    for j in remove:
        if values[j] > largestVal:
            largestVal = values[j]
            largestIn = j
            #print(largestIn)
    remove.remove(largestIn)
    order.append(largestIn)
order.append(remove[0])          
            
#print(values)
#print(vectors)
#print(order)
c=1
for i in order:
    print("lam_%d:"%(c),values[i])
    c+=1
c=1
for i in order:
    print("Ax_%d:"%(c),np.dot(A,vectors[:,i]),", lam_%d x_%d:"%(c,c),np.dot(values[i],vectors[:,i]))
    c+=1


#print("Ax_1:",np.dot(A,vectors[0]),", lam_1 x_1:",np.dot(values[0],vectors[0]))
#print("Ax_2:",np.dot(A,vectors[1]),", lam_2 x_2:",np.dot(values[1],vectors[1]))
#print("Ax_3:",np.dot(A,vectors[2]),", lam_3 x_3:",np.dot(values[2],vectors[2]))


lam_1: 15.785693179172181
lam_2: 1.2659532728352207
lam_3: -2.0516464520074047
Ax_1: [ -6.1962015   -7.01266016 -12.71289871] , lam_1 x_1: [ -6.1962015   -7.01266016 -12.71289871]
Ax_2: [ 0.12430099 -0.98053324  0.79103825] , lam_2 x_2: [ 0.12430099 -0.98053324  0.79103825]
Ax_3: [ 1.83806756 -0.74382849 -0.52676368] , lam_3 x_3: [ 1.83806756 -0.74382849 -0.52676368]


# Q4b

Without using NumPy's `la.eig()` (or similar functions), find the 4 eigenvalues and eigenvectors of the following matrix. Hint: Use the power method for the max eigenvalue (given below in `MaxEigenvalue`), the power method applied to the inverse for the smallest eigenvalue, and apply Hotelling's deflation twice to find the second and third eigenvalues/eigenvectors.

In [36]:
A = np.array([[1.935052353404, 0.887037260957, 1.534235483277, 1.370691224125],
       [0.887037260957, 0.81275989158 , 0.610238622918, 0.999150521052],
       [1.534235483277, 0.610238622918, 1.562260333538, 1.007843062517],
       [1.370691224125, 0.999150521052, 1.007843062517, 1.813258814483]])

In [64]:
def MaxEigenvalue(A, err):
    (m,n) = A.shape
    if(m != n):
        print("Matrix must be square")
        return
    x = np.random.rand(m)
    
    lam = 0.1
    lamprev = 1
    while np.abs(1-lam/lamprev) > err:
        p = np.dot(A,x)
        lamprev = lam
        lam = np.max(p)
        x = p/lam

    x = x/la.norm(lam)
    return (lam,x)

In [ ]:
eigenList = [0,0,0,0]

#maximum eigenvalue
eigenList[0] = MaxEigenvalue(A,10**-12)
sum = 0
for i in eigenList[0][1]:
    sum+= i**2

eigenList[0] = (eigenList[0][0], np.array(eigenList[0][1]/np.sqrt(sum)))

#2nd biggest eigenvalue
matrix_eig_1 = A - np.dot(eigenList[0][0],(np.outer((eigenList[0][1]),eigenList[0][1])))
eigenList[1] = MaxEigenvalue(matrix_eig_1,10**-12)

sum = 0
for i in eigenList[1][1]:
    sum+= i**2
    
eigenList[1] = (eigenList[1][0], np.array(eigenList[1][1]/np.sqrt(sum)))


#minimum eigenvalue
matrix_eig_3 = A - eigenList[0][0] * np.identity(4)
MaxEigenvalue(matrix_eig_3,10**-12)



Check your answers against Numpy's results from `la.eig(A)`

In [25]:
la.eig(A)

(array([4.91480442, 0.85058153, 0.14281707, 0.21512836]),
 array([[ 0.59947641,  0.30724064, -0.63319959, -0.3811686 ],
        [ 0.33318806, -0.3894447 ,  0.53214542, -0.67389895],
        [ 0.49528112,  0.58292464,  0.54707607,  0.34000478],
        [ 0.53320761, -0.64353313, -0.12879172,  0.53382343]]))